// @aws-cdk/aws-appmesh/lib/gateway-route-spec.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.GatewayRouteSpec = exports.GatewayRouteHostnameMatch = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const http_route_path_match_1 = require("./http-route-path-match");
const utils_1 = require("./private/utils");
const shared_interfaces_1 = require("./shared-interfaces");
/**
 * Used to generate host name matching methods.
 *
 * @stability stable
 */
class GatewayRouteHostnameMatch {
    /**
     * The value of the host name must match the specified value exactly.
     *
     * @param name The exact host name to match on.
     * @stability stable
     */
    static exactly(name) {
        return new GatewayRouteHostnameMatchImpl({ exact: name });
    }
    /**
     * The value of the host name with the given name must end with the specified characters.
     *
     * @param suffix The specified ending characters of the host name to match on.
     * @stability stable
     */
    static endsWith(suffix) {
        return new GatewayRouteHostnameMatchImpl({ suffix });
    }
}
exports.GatewayRouteHostnameMatch = GatewayRouteHostnameMatch;
_a = JSII_RTTI_SYMBOL_1;
GatewayRouteHostnameMatch[_a] = { fqn: "@aws-cdk/aws-appmesh.GatewayRouteHostnameMatch", version: "1.119.0" };
class GatewayRouteHostnameMatchImpl extends GatewayRouteHostnameMatch {
    constructor(matchProperty) {
        super();
        this.matchProperty = matchProperty;
    }
    bind(_scope) {
        return {
            hostnameMatch: this.matchProperty,
        };
    }
}
/**
 * Used to generate specs with different protocols for a GatewayRoute.
 *
 * @stability stable
 */
class GatewayRouteSpec {
    /**
     * Creates an HTTP Based GatewayRoute.
     *
     * @param options - no http gateway route.
     * @stability stable
     */
    static http(options) {
        return new HttpGatewayRouteSpec(options, shared_interfaces_1.Protocol.HTTP);
    }
    /**
     * Creates an HTTP2 Based GatewayRoute.
     *
     * @param options - no http2 gateway route.
     * @stability stable
     */
    static http2(options) {
        return new HttpGatewayRouteSpec(options, shared_interfaces_1.Protocol.HTTP2);
    }
    /**
     * Creates an gRPC Based GatewayRoute.
     *
     * @param options - no grpc gateway route.
     * @stability stable
     */
    static grpc(options) {
        return new GrpcGatewayRouteSpec(options);
    }
}
exports.GatewayRouteSpec = GatewayRouteSpec;
_b = JSII_RTTI_SYMBOL_1;
GatewayRouteSpec[_b] = { fqn: "@aws-cdk/aws-appmesh.GatewayRouteSpec", version: "1.119.0" };
class HttpGatewayRouteSpec extends GatewayRouteSpec {
    constructor(options, protocol) {
        super();
        this.routeTarget = options.routeTarget;
        this.routeType = protocol;
        this.match = options.match;
    }
    bind(scope) {
        var _c, _d, _e, _f, _g, _h, _j, _k, _l, _m;
        const pathMatchConfig = ((_d = (_c = this.match) === null || _c === void 0 ? void 0 : _c.path) !== null && _d !== void 0 ? _d : http_route_path_match_1.HttpGatewayRoutePathMatch.startsWith('/')).bind(scope);
        const rewriteRequestHostname = (_e = this.match) === null || _e === void 0 ? void 0 : _e.rewriteRequestHostname;
        const prefixPathRewrite = pathMatchConfig.prefixPathRewrite;
        const wholePathRewrite = pathMatchConfig.wholePathRewrite;
        const httpConfig = {
            match: {
                prefix: pathMatchConfig.prefixPathMatch,
                path: pathMatchConfig.wholePathMatch,
                hostname: (_g = (_f = this.match) === null || _f === void 0 ? void 0 : _f.hostname) === null || _g === void 0 ? void 0 : _g.bind(scope).hostnameMatch,
                method: (_h = this.match) === null || _h === void 0 ? void 0 : _h.method,
                headers: (_k = (_j = this.match) === null || _j === void 0 ? void 0 : _j.headers) === null || _k === void 0 ? void 0 : _k.map(header => header.bind(scope).headerMatch),
                queryParameters: (_m = (_l = this.match) === null || _l === void 0 ? void 0 : _l.queryParameters) === null || _m === void 0 ? void 0 : _m.map(queryParameter => queryParameter.bind(scope).queryParameterMatch),
            },
            action: {
                target: {
                    virtualService: {
                        virtualServiceName: this.routeTarget.virtualServiceName,
                    },
                },
                rewrite: rewriteRequestHostname !== undefined || prefixPathRewrite || wholePathRewrite
                    ? {
                        hostname: rewriteRequestHostname === undefined
                            ? undefined
                            : {
                                defaultTargetHostname: rewriteRequestHostname ? 'ENABLED' : 'DISABLED',
                            },
                        prefix: prefixPathRewrite,
                        path: wholePathRewrite,
                    }
                    : undefined,
            },
        };
        return {
            httpSpecConfig: this.routeType === shared_interfaces_1.Protocol.HTTP ? httpConfig : undefined,
            http2SpecConfig: this.routeType === shared_interfaces_1.Protocol.HTTP2 ? httpConfig : undefined,
        };
    }
}
class GrpcGatewayRouteSpec extends GatewayRouteSpec {
    constructor(options) {
        super();
        this.match = options.match;
        this.routeTarget = options.routeTarget;
    }
    bind(scope) {
        var _c;
        const metadataMatch = this.match.metadata;
        utils_1.validateGrpcGatewayRouteMatch(this.match);
        utils_1.validateGrpcMatchArrayLength(metadataMatch);
        return {
            grpcSpecConfig: {
                match: {
                    serviceName: this.match.serviceName,
                    hostname: (_c = this.match.hostname) === null || _c === void 0 ? void 0 : _c.bind(scope).hostnameMatch,
                    metadata: metadataMatch === null || metadataMatch === void 0 ? void 0 : metadataMatch.map(metadata => metadata.bind(scope).headerMatch),
                },
                action: {
                    target: {
                        virtualService: {
                            virtualServiceName: this.routeTarget.virtualServiceName,
                        },
                    },
                    rewrite: this.match.rewriteRequestHostname === undefined
                        ? undefined
                        : {
                            hostname: {
                                defaultTargetHostname: this.match.rewriteRequestHostname ? 'ENABLED' : 'DISABLED',
                            },
                        },
                },
            },
        };
    }
}
