// @aws-cdk/aws-appmesh/lib/query-parameter-match.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.QueryParameterMatch = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
/**
 * Used to generate query parameter matching methods.
 *
 * @stability stable
 */
class QueryParameterMatch {
    /**
     * The value of the query parameter with the given name in the request must match the specified value exactly.
     *
     * @param queryParameterName the name of the query parameter to match against.
     * @param queryParameterValue The exact value to test against.
     * @stability stable
     */
    static valueIs(queryParameterName, queryParameterValue) {
        return new QueryParameterMatchImpl(queryParameterName, { exact: queryParameterValue });
    }
}
exports.QueryParameterMatch = QueryParameterMatch;
_a = JSII_RTTI_SYMBOL_1;
QueryParameterMatch[_a] = { fqn: "@aws-cdk/aws-appmesh.QueryParameterMatch", version: "1.119.0" };
class QueryParameterMatchImpl extends QueryParameterMatch {
    constructor(queryParameterName, matchProperty) {
        super();
        this.queryParameterName = queryParameterName;
        this.matchProperty = matchProperty;
    }
    bind(_scope) {
        return {
            queryParameterMatch: {
                match: this.matchProperty,
                name: this.queryParameterName,
            },
        };
    }
}
