// @aws-cdk-containers/ecs-service-extensions/lib/extensions/cloudwatch-agent.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CloudwatchAgentExtension = void 0;
const ecs = require("@aws-cdk/aws-ecs");
const iam = require("@aws-cdk/aws-iam");
const extension_interfaces_1 = require("./extension-interfaces");
const CLOUDWATCH_AGENT_IMAGE = 'amazon/cloudwatch-agent:latest';
/**
 * This extension adds a CloudWatch agent to the task definition and
 * configures the task to be able to publish metrics to CloudWatch.
 */
class CloudwatchAgentExtension extends extension_interfaces_1.ServiceExtension {
    constructor() {
        super('cloudwatchAgent');
        this.CW_CONFIG_CONTENT = {
            logs: {
                metrics_collected: {
                    emf: {},
                },
            },
            metrics: {
                metrics_collected: {
                    statsd: {},
                },
            },
        };
    }
    prehook(service, scope) {
        this.parentService = service;
        this.scope = scope;
    }
    useTaskDefinition(taskDefinition) {
        // Add the CloudWatch Agent to this task
        this.container = taskDefinition.addContainer('cloudwatch-agent', {
            image: ecs.ContainerImage.fromRegistry(CLOUDWATCH_AGENT_IMAGE),
            environment: {
                CW_CONFIG_CONTENT: JSON.stringify(this.CW_CONFIG_CONTENT),
            },
            logging: new ecs.AwsLogDriver({ streamPrefix: 'cloudwatch-agent' }),
            user: '0:1338',
            memoryReservationMiB: 50,
        });
        // Add permissions that allow the cloudwatch agent to publish metrics
        new iam.Policy(this.scope, `${this.parentService.id}-publish-metrics`, {
            roles: [taskDefinition.taskRole],
            statements: [
                new iam.PolicyStatement({
                    resources: ['*'],
                    actions: ['cloudwatch:PutMetricData'],
                }),
            ],
        });
    }
    resolveContainerDependencies() {
        if (!this.container) {
            throw new Error('The container dependency hook was called before the container was created');
        }
        const appmeshextension = this.parentService.serviceDescription.get('appmesh');
        if (appmeshextension && appmeshextension.container) {
            this.container.addContainerDependencies({
                container: appmeshextension.container,
                condition: ecs.ContainerDependencyCondition.HEALTHY,
            });
        }
    }
}
exports.CloudwatchAgentExtension = CloudwatchAgentExtension;
