// @aws-cdk-containers/ecs-service-extensions/lib/extensions/extension-interfaces.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ContainerMutatingHook = exports.ServiceExtension = exports.EnvironmentCapacityType = void 0;
/**
 * The types of capacity that are supported. These capacity types may change the
 * behavior of an extension.
 */
var EnvironmentCapacityType;
(function (EnvironmentCapacityType) {
    /**
     * Specify that the environment should use AWS Fargate for
     * hosting containers.
     */
    EnvironmentCapacityType["FARGATE"] = "fargate";
    /**
     * Specify that the environment should launch containers onto
     * EC2 instances.
     */
    EnvironmentCapacityType["EC2"] = "ec2";
})(EnvironmentCapacityType = exports.EnvironmentCapacityType || (exports.EnvironmentCapacityType = {}));
/**
 * The shape of a service extension. This abstract class is implemented
 * by other extensions that extend the hooks to implement any custom
 * logic that they want to run during each step of preparing the service.
 */
class ServiceExtension {
    constructor(name) {
        // A list of other extensions which want to mutate the
        // container definition for this extension.
        this.containerMutatingHooks = [];
        this.name = name;
    }
    /**
     * A hook that allows the extension to add hooks to other
     * extensions that are registered.
     */
    addHooks() { } // tslint:disable-line
    /**
     * This hook allows another service extension to register a mutating hook for
     * changing the primary container of this extension. This is primarily used
     * for the application extension. For example, the Firelens extension wants to
     * be able to modify the settings of the application container to
     * route logs through Firelens.
     *
     * @param hook
     */
    addContainerMutatingHook(hook) {
        this.containerMutatingHooks.push(hook);
    }
    /**
     * This is a hook which allows extensions to modify the settings of the
     * task definition prior to it being created. For example, the App Mesh
     * extension needs to configure an Envoy proxy in the task definition,
     * or the Application extension wants to set the overall resource for
     * the task.
     *
     * @param props - Properties of the task definition to be created
     */
    modifyTaskDefinitionProps(props) {
        return {
            ...props,
        };
    }
    /**
     * A hook that is called for each extension ahead of time to
     * allow for any initial setup, such as creating resources in
     * advance.
     *
     * @param parent - The parent service which this extension has been added to
     * @param scope - The scope that this extension should create resources in
     */
    prehook(parent, scope) {
        this.parentService = parent;
        this.scope = scope;
    }
    /**
     * Once the task definition is created, this hook is called for each
     * extension to give it a chance to add containers to the task definition,
     * change the task definition's role to add permissions, etc.
     *
     * @param taskDefinition - The created task definition to add containers to
     */
    useTaskDefinition(taskDefinition) {
        taskDefinition = taskDefinition;
    }
    /**
     * Once all containers are added to the task definition, this hook is
     * called for each extension to give it a chance to resolve its dependency
     * graph so that its container starts in the right order based on the
     * other extensions that were enabled.
     */
    resolveContainerDependencies() {
        return;
    }
    /**
     * Prior to launching the task definition as a service, this hook
     * is called on each extension to give it a chance to mutate the properties
     * of the service to be created.
     *
     * @param props - The service properties to mutate.
     */
    modifyServiceProps(props) {
        return {
            ...props,
        };
    }
    /**
     * When this hook is implemented by extension, it allows the extension
     * to use the service which has been created. It is generally used to
     * create any final resources which might depend on the service itself.
     *
     * @param service - The generated service.
     */
    useService(service) {
        service = service;
    }
    /**
     * This hook allows the extension to establish a connection to
     * extensions from another service. Usually used for things like
     * allowing one service to talk to the load balancer or service mesh
     * proxy for another service.
     *
     * @param service - The other service to connect to.
     */
    connectToService(service) {
        service = service;
    }
}
exports.ServiceExtension = ServiceExtension;
/**
 * This is an abstract class wrapper for a mutating hook. It is
 * extended by any extension which wants to mutate other extension's containers.
 */
class ContainerMutatingHook {
    /**
     * This is a hook for modifying the container definition of any upstream
     * containers. This is primarily used for the main application container.
     * For example, the Firelens extension wants to be able to modify the logging
     * settings of the application container.
     *
     * @param props - The container definition to mutate.
     */
    mutateContainerDefinition(props) {
        return {
            ...props,
        };
    }
}
exports.ContainerMutatingHook = ContainerMutatingHook;
