// @aws-cdk-containers/ecs-service-extensions/lib/extensions/http-load-balancer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HttpLoadBalancerExtension = void 0;
const alb = require("@aws-cdk/aws-elasticloadbalancingv2");
const cdk = require("@aws-cdk/core");
const extension_interfaces_1 = require("./extension-interfaces");
/**
 * This extension add a public facing load balancer for sending traffic
 * to one or more replicas of the application container.
 */
class HttpLoadBalancerExtension extends extension_interfaces_1.ServiceExtension {
    constructor() {
        super('load-balancer');
    }
    // Before the service is created, go ahead and create the load balancer itself.
    prehook(service, scope) {
        this.parentService = service;
        this.loadBalancer = new alb.ApplicationLoadBalancer(scope, `${this.parentService.id}-load-balancer`, {
            vpc: this.parentService.vpc,
            internetFacing: true,
        });
        this.listener = this.loadBalancer.addListener(`${this.parentService.id}-listener`, {
            port: 80,
            open: true,
        });
        // Automatically create an output
        new cdk.CfnOutput(scope, `${this.parentService.id}-load-balancer-dns-output`, {
            value: this.loadBalancer.loadBalancerDnsName,
        });
    }
    // Minor service configuration tweaks to work better with a load balancer
    modifyServiceProps(props) {
        return {
            ...props,
            // Give the task a little bit of grace time to start passing
            // healthchecks. Without this it is possible for a slow starting task
            // to cause the ALB to consider the task unhealthy, causing ECS to stop
            // the task before it actually has a chance to finish starting up
            healthCheckGracePeriod: cdk.Duration.minutes(1),
        };
    }
    // After the service is created add the service to the load balancer's listener
    useService(service) {
        this.listener.addTargets(this.parentService.id, {
            deregistrationDelay: cdk.Duration.seconds(10),
            port: 80,
            targets: [service],
        });
    }
}
exports.HttpLoadBalancerExtension = HttpLoadBalancerExtension;
