// @aws-cdk-containers/ecs-service-extensions/lib/extensions/xray.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.XRayExtension = void 0;
const ecs = require("@aws-cdk/aws-ecs");
const iam = require("@aws-cdk/aws-iam");
const cdk = require("@aws-cdk/core");
const extension_interfaces_1 = require("./extension-interfaces");
const XRAY_DAEMON_IMAGE = 'amazon/aws-xray-daemon:latest';
/**
 * This extension adds an X-Ray daemon inside the task definition for
 * capturing application trace spans and submitting them to the AWS
 * X-Ray service.
 */
class XRayExtension extends extension_interfaces_1.ServiceExtension {
    constructor() {
        super('xray');
    }
    // @ts-ignore - Ignore unused params that are required for abstract class extend
    prehook(service, scope) {
        this.parentService = service;
    }
    useTaskDefinition(taskDefinition) {
        // Add the XRay Daemon to the task
        this.container = taskDefinition.addContainer('xray', {
            image: ecs.ContainerImage.fromRegistry(XRAY_DAEMON_IMAGE),
            essential: true,
            memoryReservationMiB: 256,
            environment: {
                AWS_REGION: cdk.Stack.of(this.parentService).region,
            },
            healthCheck: {
                command: [
                    'CMD-SHELL',
                    'curl -s http://localhost:2000',
                ],
                startPeriod: cdk.Duration.seconds(10),
                interval: cdk.Duration.seconds(5),
                timeout: cdk.Duration.seconds(2),
                retries: 3,
            },
            logging: new ecs.AwsLogDriver({ streamPrefix: 'xray' }),
            user: '1337',
        });
        // Add permissions to this task to allow it to talk to X-Ray
        taskDefinition.taskRole.addManagedPolicy(iam.ManagedPolicy.fromAwsManagedPolicyName('AWSXRayDaemonWriteAccess'));
    }
    resolveContainerDependencies() {
        if (!this.container) {
            throw new Error('The container dependency hook was called before the container was created');
        }
        const appmeshextension = this.parentService.serviceDescription.get('appmesh');
        if (appmeshextension && appmeshextension.container) {
            this.container.addContainerDependencies({
                container: appmeshextension.container,
                condition: ecs.ContainerDependencyCondition.HEALTHY,
            });
        }
    }
}
exports.XRayExtension = XRayExtension;
