// @aws-cdk-containers/ecs-service-extensions/lib/extensions/assign-public-ip/assign-public-ip.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AssignPublicIpExtension = void 0;
const ec2 = require("@aws-cdk/aws-ec2");
const extension_interfaces_1 = require("../extension-interfaces");
const task_record_manager_1 = require("./task-record-manager");
/**
 * Modifies the service to assign a public ip to each task and optionally
 * exposes public IPs in a Route 53 record set.
 *
 * Note: If you want to change the DNS zone or record name, you will need to
 * remove this extension completely and then re-add it.
 */
class AssignPublicIpExtension extends extension_interfaces_1.ServiceExtension {
    constructor(options) {
        super('public-ip');
        this.dns = options === null || options === void 0 ? void 0 : options.dns;
    }
    hasDns() {
        return Boolean(this.dns);
    }
    prehook(service, _scope) {
        super.prehook(service, _scope);
        if (service.capacityType != extension_interfaces_1.EnvironmentCapacityType.FARGATE) {
            throw new Error('AssignPublicIp only supports Fargate tasks');
        }
    }
    modifyServiceProps(props) {
        return {
            ...props,
            assignPublicIp: true,
        };
    }
    useService(service) {
        if (this.hasDns()) {
            new task_record_manager_1.TaskRecordManager(service, 'TaskRecordManager', {
                service: service,
                dnsZone: this.dns.zone,
                dnsRecordName: this.dns.recordName,
            });
            const container = this.parentService.serviceDescription.get('service-container');
            service.connections.allowFromAnyIpv4(ec2.Port.tcp(container.trafficPort), 'Accept inbound traffic on traffic port from anywhere');
        }
    }
}
exports.AssignPublicIpExtension = AssignPublicIpExtension;
